solr/core/src/java/org/apache/solr/handler/component/SlowNodeDetector.java

```diff
@@ -144,11 +148,67 @@ private Set<String> computeSlowNodes(RequestStats stats) {
       }
     }
 
-    if (log.isInfoEnabled() && !slowNodes.isEmpty()) {
-      log.info("Slow nodes detected: {}", slowNodes);
+    // find previous slow nodes that might have recovered
+    Set<String> recoveredNodes = new HashSet<>();
+    Set<String> recoveredSlowNodeCandidates = new HashSet<>(this.slowNodes.keySet());
+    recoveredSlowNodeCandidates.removeAll(slowNodes);
+
+    if (!recoveredSlowNodeCandidates.isEmpty()) {
+      // the threshold for sorted response to be considered as normal latency
+      int normalNodeResponseCount = stats.responseLatencies.size() - maxSlowResponseCount;
+      iteratedResponseCountByNode.clear();
+      for (int i = stats.responseLatencies.size() - 1; i >= normalNodeResponseCount; i--) {
+        RequestStats.NodeLatency latency = stats.responseLatencies.get(i);
+        if (!recoveredSlowNodeCandidates.contains(latency.node)) {
+          continue;
+        }
+        iteratedResponseCountByNode.compute(latency.node, (k, v) -> v == null ? 1 : v + 1);
```

Comment:
I missed this before, but for what we're doing with `iteratedResponseCountByNode`, `com.carrotsearch.hppc.ObjectIntHashMap` would be a great alternative (specifically, the `putOrAdd()` method. Orthogonal/not relevant to the current PR though!